modules/fcrepo/main.tf

```diff
@@ -0,0 +1,150 @@
+# Actual direct beanstalk configuration for the fedora core.
+resource "aws_elastic_beanstalk_application" "fedora" {
+  name        = "${var.StackName}"
+  description = "Fcrepo service"
+}
+
+resource "aws_elastic_beanstalk_environment" "fedora" {
+  name          = "${var.StackName}"
+  description   = "Fcrepo Environment"
+  application   = "${aws_elastic_beanstalk_application.fedora.name}"
+  template_name = "${aws_elastic_beanstalk_configuration_template.fedora.name}"
+  version_label = "${aws_elastic_beanstalk_application_version.fedora.name}"
+}
+
+resource "aws_elastic_beanstalk_application_version" "fedora" {
+  name        = "fcrepo-version-label"
+  application = "${aws_elastic_beanstalk_application.fedora.name}"
+  description = "Fcrepo service version"
+  bucket      = "${var.S3Bucket}"
+  key         = "${var.S3Key}"
+}
```

Comment:
Putting the application versions in the module makes it hard to reuse unless you make a separate app version for each instance of the module

Reply:
Oh yeah, I'd need to at least embed StackName into the name to avoid that case, probably.

For pulling it out of the module, we'd also need to pull out aws_elastic_beanstalk_application, since the application_version relies on it.  Is this a case where we're getting a lot out of doing so? At least in the confines of this, it looks like we're only loading fcrepo once, so the benefit would be if we're running multiple hyrax stacks.  Is that a normal usecase?

Reply:
Within a single hyrax stack fedora will only be loaded once but we might need to install fedora multiple times for sure. We currently have I think 4 or 5 fedoras running currently in our on premise infrastructure (many in testing/dev environments). I ended up abstracting out both the beanstalk application creation and the application version functionality on their own in the terraform-aws repo. Looking back at it, I think the application module is overkill but the application version one might be a good pattern to replicate: https://github.com/sul-dlss/terraform-aws/tree/master/modules/beanstalk_application_version

Reply:
Okay, so you were coming at this from the perspective of a generalized fedora module.  The followup question I'd have as someone not actually familiar with the application -- how much of what's in here is general fedora and how much is specific to the hyrax stack configuration?  Are there a lot of things in the module as is (aside from the app version) that we'd need to make configurable as part of it being a broader module?  I'm mostly concerned with the configuration template, I think.  How much of that would be something we'd need to make configurable that isn't already?  

It sounds like a great plan to me, I'm just wanting to make sure it's all properly done that way.

Reply:
It's an excellent question and it feels like there are lots of configuration decisions to make in order to do it right. Perhaps we could have a conservation about what we feel like should be general assumed practice and what should be configurable etc.?

Reply:
That's good here, definitely.

Though for the specific, in this case my problem is that I don't know much about fedora, and am just not sure what settings here are more specific to this stack and which are just normal.  For instance, the jvm settings.  In a general fedora module, I figure those would probably be configurable, but I'm not sure.  And if they do need to be configurable, is it a case where they could all change (and thus we'd want that entire setting to be loaded from a variable) or are they mostly always the same but we need to do an option for more (and so maybe just add a jvm_extra field to the template's last line).

Since this is I think bigger picture work, I'm going to go ahead and create a ticket to look at refactoring this module later and commit as is so that it's not a blocker for the stack in the meantime.